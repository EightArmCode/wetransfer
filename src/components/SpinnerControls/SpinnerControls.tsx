import React, { FC, ReactElement } from 'react'
import { Button, ButtonGroup, Slider } from '@material-ui/core'
import { css, StyleSheet } from 'aphrodite/no-important'

type Props = {
    active:boolean
    progress:number
    resetIntervalRefs:() => void
    setProgress:(progress:number) => void
    setVelocity:(velocity:number) => void
    setActive:(active:boolean) => void
    velocity:number
}

const SpinnerControls:FC<Props> = ({
    
    active,
    setActive,
    
    progress,
    setProgress,
    
    velocity,
    setVelocity,
    
    resetIntervalRefs,

}):ReactElement => {
    
    const marks = [
		{
			value: 5,
			label: '5%',
		},
		{
			value: 10,
			label: '10%',
		},
		{
			value: 25,
			label: '25%',
		},
		{
			value: 50,
			label: '50%',
		},
		{
			value: 75,
			label: '75%',
		},
		{
			value: 90,
			label: '90%',
		},
	]

    return (
        <div className={css(styles.interactions)}>
			<div className={css(styles.buttonsWrapper)}>

                <ButtonGroup color="primary" aria-label="outlined primary button group">
                    
                    <Button
                        disabled={progress >= 1}
                        id="start"
                        onClick={() => setActive(true)}
                    >
                        Start
                    </Button>

                    <Button
                        id="stop"
                        onClick={() => {
                            setActive(false)
                            resetIntervalRefs()
                        }}
                    >
                        Stop
                    </Button>
                    
                    <Button
                        disabled={active}
                        id="reset"
                        onClick={() => {
                            resetIntervalRefs()
                            setProgress(0)
                        }}
                    >
                        Reset
                    </Button>
                </ButtonGroup>
            </div>
            <Slider
                aria-labelledby="discrete-slider-custom"
                getAriaValueText={(value:number) => `${value}%`}
                id="slider"
                marks={marks}
                onChange={(e, value) => {
                    // do not accept array value
                    if (!Array.isArray(value)) setVelocity(value)
                }}
                role="slider"
                step={null}
                data-testid="internal-slider"
                value={velocity}
                valueLabelDisplay="auto"
            />
        </div>
    )
}

const styles = StyleSheet.create({
    interactions: {
		justifyContent: 'center',
		display: 'flex',
		flexDirection: 'column',
		backgroundColor: '#eaf2dd',
		// marginTop: '5%',
		padding: 24,
	},
	buttonsWrapper: {
		display: 'flex',
		justifyContent: 'center',
		marginBottom: 24,
	}
})

export default SpinnerControls