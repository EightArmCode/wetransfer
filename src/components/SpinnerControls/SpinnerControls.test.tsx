import React from 'react'
import SpinnerControls from './SpinnerControls'
import {cleanup, render, fireEvent} from '@testing-library/react'
import Enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'

Enzyme.configure({ adapter: new Adapter() })

let mockFn: () => void
beforeEach(() => {
    mockFn = jest.fn()
})

afterEach(cleanup)

describe('<SpinnerControls />', () => {
    it('renders correctly', () => {
        const { getByRole, getAllByRole, getByTestId } = render(
            <SpinnerControls
                active={false}
                progress={0}
                setActive={mockFn}
                setProgress={mockFn}
                setVelocity={mockFn}
                velocity={5}
                resetIntervalRefs={mockFn}
            />
        )

        const buttonGroup = getByRole('group')
        expect(buttonGroup).toBeDefined()
        expect(buttonGroup.children.length).toEqual(3)

        // There is an element assigned the slider role by the material UI team,
        // and one added in our code
        const sliders = getAllByRole('slider')
        expect(sliders.length).toEqual(2)

        const internalSlider = getByTestId('internal-slider')
        expect(internalSlider).toBeDefined()
    })

    it('responds correctly to button clicks', async () => {
        const mockFn = jest.fn()
        const { getByRole, container, rerender, findByRole } = render(
            <SpinnerControls
                active={false}
                progress={0}
                setActive={mockFn}
                setProgress={mockFn}
                setVelocity={mockFn}
                velocity={5}
                resetIntervalRefs={mockFn}
            />
        )
    
        expect(container).toMatchSnapshot()
    
        rerender(
            <SpinnerControls
                active={true}
                progress={0}
                setActive={mockFn}
                setProgress={mockFn}
                setVelocity={mockFn}
                velocity={5}
                resetIntervalRefs={mockFn}
            />
        )
    
        fireEvent.click(getByRole('button', { name: 'Start'}))

        const reset = await findByRole('button', { name: 'Reset' })
        expect(reset).toBeDisabled()
    
        rerender(
            <SpinnerControls
                active={false}
                progress={1}
                setActive={mockFn}
                setProgress={mockFn}
                setVelocity={mockFn}
                velocity={5}
                resetIntervalRefs={mockFn}
            />
        )

        expect(getByRole('button', { name: 'Start' })).toBeDisabled()
    
        fireEvent.click(getByRole('button', { name: 'Reset' }))

        rerender(
            <SpinnerControls
                active={false}
                progress={0}
                setActive={mockFn}
                setProgress={mockFn}
                setVelocity={mockFn}
                velocity={5}
                resetIntervalRefs={mockFn}
            />
        )

        expect(getByRole('button', { name: 'Start' })).not.toBeDisabled()
    
        expect(container).toMatchSnapshot()
    })
})