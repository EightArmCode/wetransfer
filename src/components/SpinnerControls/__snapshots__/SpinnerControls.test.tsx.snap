// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpinnerControls /> responds correctly to button clicks 1`] = `
.interactions {
  -webkit-box-direction: normal;
  -webkit-box-orient: vertical;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #eaf2dd;
  padding: 24px;
}

.buttonsWrapper {
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: center;
  justify-content: center;
  margin-bottom: 24px;
}

<div>
  <div
    class="interactions"
  >
    <div
      class="buttonsWrapper"
    >
      <div
        aria-label="outlined primary button group"
        class="MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="start"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Start
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="stop"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Stop
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="reset"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reset
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <span
      class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-marked"
      data-testid="internal-slider"
      id="slider"
      role="slider"
    >
      <span
        class="MuiSlider-rail"
      />
      <span
        class="MuiSlider-track"
        style="left: 0%; width: 5%;"
      />
      <input
        type="hidden"
        value="5"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="0"
        style="left: 5%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel MuiSlider-markLabelActive"
        data-index="0"
        style="left: 5%;"
      >
        5%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="1"
        style="left: 10%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="1"
        style="left: 10%;"
      >
        10%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="2"
        style="left: 25%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="2"
        style="left: 25%;"
      >
        25%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="3"
        style="left: 50%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="3"
        style="left: 50%;"
      >
        50%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="4"
        style="left: 75%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="4"
        style="left: 75%;"
      >
        75%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="5"
        style="left: 90%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="5"
        style="left: 90%;"
      >
        90%
      </span>
      <span
        aria-labelledby="discrete-slider-custom"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="5"
        aria-valuetext="5%"
        class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-6"
        data-index="0"
        role="slider"
        style="left: 5%;"
        tabindex="0"
      >
        <span
          class="PrivateValueLabel-offset-8 MuiSlider-valueLabel"
        >
          <span
            class="PrivateValueLabel-circle-9"
          >
            <span
              class="PrivateValueLabel-label-10"
            >
              5
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</div>
`;

exports[`<SpinnerControls /> responds correctly to button clicks 2`] = `
<div>
  <div
    class="interactions_en6gbv"
  >
    <div
      class="buttonsWrapper_1y763qn"
    >
      <div
        aria-label="outlined primary button group"
        class="MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="start"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Start
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="stop"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Stop
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
          id="reset"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reset
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <span
      class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-marked"
      data-testid="internal-slider"
      id="slider"
      role="slider"
    >
      <span
        class="MuiSlider-rail"
      />
      <span
        class="MuiSlider-track"
        style="left: 0%; width: 5%;"
      />
      <input
        type="hidden"
        value="5"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="0"
        style="left: 5%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel MuiSlider-markLabelActive"
        data-index="0"
        style="left: 5%;"
      >
        5%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="1"
        style="left: 10%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="1"
        style="left: 10%;"
      >
        10%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="2"
        style="left: 25%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="2"
        style="left: 25%;"
      >
        25%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="3"
        style="left: 50%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="3"
        style="left: 50%;"
      >
        50%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="4"
        style="left: 75%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="4"
        style="left: 75%;"
      >
        75%
      </span>
      <span
        class="MuiSlider-mark"
        data-index="5"
        style="left: 90%;"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-markLabel"
        data-index="5"
        style="left: 90%;"
      >
        90%
      </span>
      <span
        aria-labelledby="discrete-slider-custom"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="5"
        aria-valuetext="5%"
        class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-6"
        data-index="0"
        role="slider"
        style="left: 5%;"
        tabindex="0"
      >
        <span
          class="PrivateValueLabel-offset-8 MuiSlider-valueLabel"
        >
          <span
            class="PrivateValueLabel-circle-9"
          >
            <span
              class="PrivateValueLabel-label-10"
            >
              5
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</div>
`;
